velox/functions/sparksql/ToPrettyString.h

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/expression/CastExpr-inl.h"
+#include "velox/functions/Udf.h"
+#include "velox/type/Conversions.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace detail {
+static const StringView kNull = "NULL";
+}
+
+template <typename TExec>
+struct ToPrettyStringFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  template <typename A>
+  void initialize(
+      const std::vector<TypePtr>& inputTypes,
+      const core::QueryConfig& /*config*/,
+      A* /*a*/) {
+    inputType_ = inputTypes[0];
+  }
+
+  template <typename TInput>
+  void call(out_type<Varchar>& result, const TInput& input) {
+    if constexpr (std::is_same_v<TInput, StringView>) {
+      if (inputType_->isVarchar()) {
+        result = input;
+      } else if (inputType_->isVarbinary()) {
+        result.resize(1 + 3 * input.size());
+        char* const startPosition = result.data();
+        char* pos = startPosition;
+        *pos++ = '[';
+        for (auto i = 0; i < input.size(); i++) {
+          auto formated = fmt::format("{:X}", input.data()[i]);
+          *pos++ = formated.data()[0];
+          *pos++ = formated.data()[1];
+          *pos++ = ' ';
+        }
+        *--pos = ']';
+      }
+      return;
+    }
+    if constexpr (std::is_same_v<TInput, int32_t>) {
+      if (inputType_->isDate()) {
+        auto output = DATE()->toString(input);
+        result.append(output);
+        return;
+      }
+    }
+    const auto castResult =
+        util::Converter<TypeKind::VARCHAR, void, util::SparkCastPolicy>::
+            tryCast(input);
+    if (castResult.hasError()) {
+      result.setNoCopy(detail::kNull);
+    } else {
+      result.copy_from(castResult.value());
+    }
+  }
+
+  template <typename TInput>
+  void callNullable(out_type<Varchar>& result, const TInput* a) {
+    if (a) {
+      call(result, *a);
+    } else {
+      result.setNoCopy(detail::kNull);
+    }
+  }
+
+ private:
+  TypePtr inputType_;
+};
+
+template <typename TExec>
+struct ToPrettyStringTimeStampFunction {
```

Comment:
Could you please explain the differences between `ToPrettyStringTimeStampFunction` & `ToPrettyStringDecimalFunction` with their cast implementations? I notice Spark uses `castToString` for all types on non-null values, and can we find a better way to reuse the cast implementations? Thanks.

Reply:
It is same for non-null value, if we need to do that,  we should do refactor for current cast implementation, and this is simple function, cast is for vector, it is different.

Maybe we can change this one to vector function.
And do the refactor in another PR or this PR, I'm open for these 2 ways.

Reply:
I notice CastExpr has an exposed API `evalSpecialForm`. Does it make sense to reuse the implementation of cast with it? @mbasmanova Do you have any suggestion? Thanks.

Reply:
We should also construct `CastExpr` by 
```
  CastExpr(
      TypePtr type,
      ExprPtr&& expr,
      bool trackCpuUsage,
      bool nullOnFailure,
      std::shared_ptr<CastHooks> hooks)
      : SpecialForm(
            type,
            std::vector<ExprPtr>({expr}),
            nullOnFailure ? kTryCast.data() : kCast.data(),
            false /* supportsFlatNoNullsFastPath */,
            trackCpuUsage),
        nullOnFailure_(nullOnFailure),
        hooks_(std::move(hooks)) {}
```

We don't have `ExprPtr` here.

Reply:
binary logic is different with cast for non-null value

Reply:
It is not possible to invoke CAST expression from inside a function. The only way to re-use it is to re-use underlying logic. Alternatively, you may re-write the expression from toprettystring(x) as CAST(x as varchar), but it looks like this will not produce the right outcome.

I'll be on vacation from tomorrow till July 15.

Reply:
@mbasmanova Thanks for the pointer. I understand. Enjoy the vacation :)